controllers/virtualmachinegrouppublishrequest/virtualmachinegrouppublishrequest_controller.go

```diff
@@ -95,63 +127,308 @@ func NewReconciler(
+	}
+
+	existReqsMap, pendingReqsSet, completedReqsSet, err := r.getVMPublishRequests(ctx)
+	if err != nil {
+		return err
+	}
+
+	r.reconcileStatusImages(ctx, existReqsMap)
+
+	desiredReqsSet := sets.New(ctx.VMGroupPublishRequest.Spec.VirtualMachines...)
+	if err := r.reconcileNewPublishRequests(ctx, sets.KeySet(existReqsMap), desiredReqsSet); err != nil {
+		return err
+	}
+	return r.reconcileStatusCompletedCondition(ctx, pendingReqsSet, completedReqsSet, desiredReqsSet)
+}
+
+// getVMPublishRequests returns a map of VMGroupPublishRequests own by the VirtualMachineGroupPublishRequest
+// along with a set of pending VMGroupPublishRequest and a set of completed VMGroupPublishRequest.
+// The key for those objects is the source VirtualMachine name.
+func (r *Reconciler) getVMPublishRequests(
+	ctx *pkgctx.VirtualMachineGroupPublishRequestContext) (
+	map[string]vmopv1.VirtualMachinePublishRequest,
+	sets.Set[string],
+	sets.Set[string],
+	error) {
+
+	reqs := &vmopv1.VirtualMachinePublishRequestList{}
+	// retrieve from API server directly to avoid duplicate resource creation in case that the newly
+	// created vm publish request has not made it back into the client's cache
+	if err := r.apiReader.List(ctx,
+		reqs,
+		client.InNamespace(ctx.VMGroupPublishRequest.Namespace),
+		client.MatchingLabels{
+			vmopv1.VirtualMachinePublishRequestManagedByLabelKey: ctx.VMGroupPublishRequest.Name,
+		}); err != nil {
+		ctx.Logger.Error(err, "failed to list VirtualMachinePublishRequest")
+		return nil, nil, nil, err
+	}
+
+	existReqsMap := make(map[string]vmopv1.VirtualMachinePublishRequest)
+	pendingReqsSet := sets.Set[string]{}
+	completedReqsSet := sets.Set[string]{}
+	for _, req := range reqs.Items {
+		if metav1.IsControlledBy(&req, ctx.VMGroupPublishRequest) {
+			existReqsMap[req.Spec.Source.Name] = req
+			if req.Status.Ready {
+				completedReqsSet.Insert(req.Spec.Source.Name)
+			} else {
+				pendingReqsSet.Insert(req.Spec.Source.Name)
+			}
+		}
+	}
+
+	return existReqsMap, pendingReqsSet, completedReqsSet, nil
+}
+
+// reconcileStatusImages updates status's image status by taking a map of VMGroupPublishRequests
+// own by the VirtualMachineGroupPublishRequest and converting them into a sorted list of
+// VirtualMachineGroupPublishRequestImageStatus. Sorting prevents reconciling when list is unchanged.
+func (r *Reconciler) reconcileStatusImages(
+	ctx *pkgctx.VirtualMachineGroupPublishRequestContext,
+	existReqsMap map[string]vmopv1.VirtualMachinePublishRequest) {
+
+	imageStatuses := make([]vmopv1.VirtualMachineGroupPublishRequestImageStatus, 0, len(existReqsMap))
+
+	for _, req := range existReqsMap {
+		imageStatuses = append(imageStatuses, vmopv1.VirtualMachineGroupPublishRequestImageStatus{
+			Source:             req.Spec.Source.Name,
+			PublishRequestName: req.Name,
+			ImageName:          req.Status.ImageName,
+			Conditions:         req.Status.Conditions,
+		})
+	}
+
+	sort.Slice(imageStatuses, func(i, j int) bool {
```

Comment:
I think in newer golang version it's recommended to use `slices.Sort/SortFunc()` for better performance.